marathon/src/main/scala/io/buoyant/marathon/v2/Api.scala

```diff
@@ -101,21 +112,90 @@ object Api {
   }
 }
 
-private class AppIdApi(client: Api.Client, apiPrefix: String)
+private class AppIdApi(client: Api.Client, apiPrefix: String, auth: Option[Api.Auth])
   extends Api
   with Closable {
 
   import Api._
 
+  var authToken: Option[String] = None
+
   def close(deadline: Time) = client.close(deadline)
 
-  def getAppIds(): Future[Api.AppIds] = {
-    val req = http.Request(s"$apiPrefix/apps")
-    Trace.letClear(client(req)).flatMap(rspToApps(_))
-  }
+  private[this] def refreshToken(): Future[String] =
+    auth match {
+      case Some(auth) =>
+        val req = http.Request(http.Method.Post, auth.path)
+        req.setContentTypeJson()
+        req.setContentString(auth.content)
+
+        Trace.letClear(client(req)).flatMap { rsp =>
+          rsp.status match {
+            case http.Status.Ok =>
+              readJson[AuthToken](rsp.content) match {
+                case Return(AuthToken(Some(token))) =>
+                  authToken = Some(token) // side effect
+                  Future.const(Return(token))
+                case _ => Future.exception(UnauthorizedResponse(rsp))
+              }
+            case _ => Future.exception(UnauthorizedResponse(rsp))
+          }
+        }
+      case _ => Future.value("")
+    }
 
-  def getAddrs(app: Path): Future[Set[Address]] = {
-    val req = http.Request(s"$apiPrefix/apps${app.show}")
-    Trace.letClear(client(req)).flatMap(rspToAddrs(_))
+  private[this] def prepRequest(path: String): Future[http.Request] = {
+    val req = http.Request(path)
+    (authToken, auth) match {
+      case (Some(authToken), Some(auth)) => {
+        // already have a token, set it
+        req.headerMap.set("Authorization", s"token=$authToken")
+        Future.const(Return(req))
+      }
+      case (None, Some(auth)) => {
+        // need a token, query for it
+        refreshToken().flatMap { token =>
+          req.headerMap.set("Authorization", s"token=$token")
+          Future.const(Return(req))
+        }
+      }
+
+      // don't need a token
+      case _ => Future.const(Return(req))
+    }
   }
+
+  def getAppIds(): Future[Api.AppIds] =
+    prepRequest(s"$apiPrefix/apps").flatMap { req =>
+      try {
+        Trace.letClear(client(req)).flatMap(rspToApps(_)).rescue {
+          case e: UnauthorizedResponse =>
+            // token may have expired, try again
+            refreshToken().flatMap { token =>
+              req.headerMap.set("Authorization", s"token=$token")
+              Trace.letClear(client(req)).flatMap(rspToApps(_))
```

Comment:
what if you get UnauthorizedResponse back from this one?  Do we assume that there's no way the token could expire in this span of time?